docs/proposals/high-availability.md

```diff
@@ -0,0 +1,34 @@
+# High Availability of Daemons in Kubernetes
+This document serves as a proposal for high availability of the master daemons in kubernetes.
+
+## Design Options
+1. Hot Standby Daemons: In this scenario, data and state are shared between the two deamons such that an immediate failure in one daemon causes the the standby deamon to take over exactly where the failed daemon had left off.  This would be an ideal solution for kubernetes, however it poses a series of challenges in the case of controllers where daemon-state is cached locally and not persisted in a transactional way to a storage facility.  As a result, we are **NOT** planning on this approach. 
+
+2. **Cold Standby Daemons**: In this scenario there is only one active daemon acting as the master and additional daemons in a standby mode.  Data and state are not shared between the active and standby daemons, so when a failure occurs the standby daemon that becomes the master must determine the current state of the system before resuming functionality.
+
+3. Stateless load-balanced Daemons: Stateless daemons, such as the apiserver, can simply load-balance across any number of servers that are currently running.  Their general availability can be continuously updated, or published, such that load balancing only occurs across active participants.  This aspect of HA is outside of the scope of *this* proposal because there is already a partial implementation in the apiserver.
+
+
+## Design Discussion Notes on Leader Election
+For a very simple example of proposed behavior see: 
+* https://github.com/rrati/etcd-ha
+* go get github.com/rrati/etcd-ha
+
+In HA, the apiserver will be a gateway to etcd. It will provide an api for becoming master, updating the master lease, and releasing the lease.  This api is daemon agnostic, so to become the master the client will need to provide the daemon type and the lease duration when attemping to become master.  The apiserver will attempt to create a key in etcd based on the daemon type that contains the client's hostname/ip and port information. This key will be created with a ttl from the lease duration provided in the request.  Failure to create this key means there is already a master of that daemon type, and the error from etcd will propigate to the client.  Successfully creating the key means the client making the request is the master.  When updating the lease, the apiserver will update the existing key with a new ttl.  The location in etcd for the HA keys is TBD.
```

Comment:
"In HA, the apiserver will be a gateway to etcd." -> Is it already the case that api server is the only gateway to etcd?

" It will provide an api for becoming master, ..." -> Suggest clarifying this sentence to "The api server will provide an api for sets of replicated clients to do master election: become msater, upate the master lease, and release the lease."